vmm/src/interrupt.rs

```diff
@@ -142,6 +148,10 @@ where
     MsiInterruptGroup<E>: MsiInterruptGroupOps,
 {
     fn enable(&self) -> Result<()> {
+        self.devid_required.store(
+            self.vm.check_extension(hypervisor::Cap::MsiDevid),
+            Ordering::SeqCst,
+        );
```

Comment:
The reason to use `AtomicBool` instead of simple `bool` for `devid_required` is: `devid_required` should be determined in `enable()`, but the parameter here is `&self`, by which I cannot modify its member. Using `AtomicBool` is a convenient option, although we don't need something atomic.

Please suggest if there is better way.

Reply:
A dumb question: why should `devid_required` be determined in `enable()`? Does the `MsiDevid` extension come and go during the life cycle of the guest?

Reply:
It could be special on Arm (not sure if it's same on X86): The requirement for `devid` was determined by GIC device (generic interrupt controller for Arm). When the group was created, GIC hadn't been created, it's impossible to know whether `devid` is needed or not.

Reply:
OK, so this is a problem with the code structure.

My observation here is that checking the extension doesn't necessarily need to be tied with the creation of the GIC in use -- you didn't need to reference the interrupt controller structure. Can you determine this property in `fn new` in `impl<E> MsiInterruptGroup<E>`? If so, that would save you the need to repeatedly store this boolean.

One further idea is to use `Option<u32>` for `devid` to get rid of this `devid_required` all together. But I'm probably getting ahead of myself here.

Anyway, just trying to help. I'm not a maintainer, so I will let you and the maintainers figure out what is best.

Reply:
The result of checking the extension depends on the creation of GIC (GIC devices emulated by KVM). The logic is in kernel. The real result of the extension was only available after GIC device created in KVM. Before that, the result was always False (default value).

I guess the "GIC" defined here `devices/src/gic.rs` is confusing you, it is something used for providing interrupt service (enable, trigger, ...). While the device is emulated here: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/arch/src/aarch64/mod.rs#L149. The extension result depends on the creation of the latter one.

`MsiInterruptGroup::new()` is too early, GIC device hasn't been created in KVM then.

Reply:
@liuw, Thanks a lot for looking into this. Hopefully my explanation helps. :)

To be simple: Checking the extension need to be done after this (called from `Vm::boot()`) https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/arch/src/aarch64/mod.rs#L149, but `MsiInterruptGroup::new()` is done before it.

Reply:
I see. Can the check be done in `make_entry` alone? I think the logic here is you will have to have `devid_required` correctly set at some point before entering `make_entry`. Since `MsiDevid` cap is persistent throughout the life cycle of the VM, it is functionally equivalent that you check the extension in `make_entry` at the last possible moment.

That means you can get rid of `devid_required` and other plumbing code all together. See https://github.com/liuw/cloud-hypervisor/commit/6a7b0510fa14bcf1d855c5548e9bc28a774505ed. Can that work?

Reply:
Thank you again for help. :) Yes, it should work, and it makes the least code change.
Making the ioctl each time in `make_entry` will take more time, but the cost should be very tiny (in my test, `make_entry` was called twice in initializing a virtio-blk device).

Reply:
I think doing it in `make_entry` and `update` will have the same degree of performance cost FWIW.

I have thought a bit about if it is possible to determine that extension feature early so that we don't need to query KVM every time a change is made, but haven't come up with a solution that doesn't require much code churn.

Anyway, I think the approach I proposed should satisfy your need for now. Feel free to squash my change to your patch.

Reply:
Code was updated, thanks.

Reply:
LGTM!